From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: kenway214 <kenway214@outlook.com>
Date: Mon, 3 Aug 2026 04:44:26 +0530
Subject: [PATCH 26/27] base: Add support for launching hidden/blocked apps
 from Obscura

Change-Id: I8a3b75e405d4255728e194a4b9bde225659f4288
(cherry picked from commit 7ea09c4de03fff8f9883bc0bdb0e9353eee93cdd)
---
 core/java/android/app/ObscuraManager.java     |  8 +++++
 .../android/internal/app/IObscuraManager.aidl |  2 ++
 .../server/obscura/AppControlController.java  | 35 +++++++++++++++++++
 .../server/obscura/IObscuraService.java       |  1 +
 .../server/obscura/ObscuraService.java        |  4 +++
 5 files changed, 50 insertions(+)

diff --git a/core/java/android/app/ObscuraManager.java b/core/java/android/app/ObscuraManager.java
index 1f3b601d5e83..dbd193623db7 100644
--- a/core/java/android/app/ObscuraManager.java
+++ b/core/java/android/app/ObscuraManager.java
@@ -246,4 +246,12 @@ public class ObscuraManager {
             throw e.rethrowFromSystemServer();
         }
     }
+
+    public void launchHiddenApp(@NonNull String packageName) {
+        try {
+            mService.launchHiddenApp(packageName);
+        } catch (RemoteException e) {
+            throw e.rethrowFromSystemServer();
+        }
+    }
 }
diff --git a/core/java/com/android/internal/app/IObscuraManager.aidl b/core/java/com/android/internal/app/IObscuraManager.aidl
index 173ca16becba..9d69144c3a4a 100644
--- a/core/java/com/android/internal/app/IObscuraManager.aidl
+++ b/core/java/com/android/internal/app/IObscuraManager.aidl
@@ -41,4 +41,6 @@ interface IObscuraManager {
     void setDataIsolationEnabled(String packageName, boolean enabled);
 
     String getSpoofedSetting(String callingPackage, String settingName);
+
+    void launchHiddenApp(String packageName);
 }
diff --git a/services/core/java/com/android/server/obscura/AppControlController.java b/services/core/java/com/android/server/obscura/AppControlController.java
index bafc77ecac27..8509497caa86 100644
--- a/services/core/java/com/android/server/obscura/AppControlController.java
+++ b/services/core/java/com/android/server/obscura/AppControlController.java
@@ -444,4 +444,39 @@ public class AppControlController {
             if (changed) saveConfigToSettings();
         }
     }
+
+    public void launchHiddenApp(String packageName) {
+        if (TextUtils.isEmpty(packageName)) return;
+        boolean wasHidden = isPackageHidden(packageName);
+        boolean wasLauncherHidden = isPackageLauncherHidden(packageName);
+
+        if (wasHidden) {
+            setPackageHidden(packageName, false);
+        }
+        if (wasLauncherHidden) {
+            setPackageLauncherHidden(packageName, false);
+        }
+
+        // Launch the app
+        try {
+            Intent intent = mContext.getPackageManager()
+                    .getLaunchIntentForPackage(packageName);
+            if (intent != null) {
+                intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK
+                        | Intent.FLAG_ACTIVITY_CLEAR_TOP
+                        | Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED);
+                mContext.startActivityAsUser(intent, UserHandle.CURRENT);
+            }
+        } catch (Exception e) {
+            Slog.e(TAG, "Failed to launch hidden app: " + packageName, e);
+        }
+
+        // Re-hide after a short delay to allow launch to complete
+        if (wasHidden || wasLauncherHidden) {
+            mHandler.postDelayed(() -> {
+                if (wasHidden) setPackageHidden(packageName, true);
+                if (wasLauncherHidden) setPackageLauncherHidden(packageName, true);
+            }, 2000);
+        }
+    }
 }
diff --git a/services/core/java/com/android/server/obscura/IObscuraService.java b/services/core/java/com/android/server/obscura/IObscuraService.java
index eb11dab583ba..87ea970e01e2 100644
--- a/services/core/java/com/android/server/obscura/IObscuraService.java
+++ b/services/core/java/com/android/server/obscura/IObscuraService.java
@@ -17,4 +17,5 @@ public interface IObscuraService {
     String getSpoofedSetting(String callingPackage, String settingName);
     int[] getRestrictedGids(String packageName);
     boolean isDataIsolationEnabled(String packageName);
+    void launchHiddenApp(String packageName);
 }
diff --git a/services/core/java/com/android/server/obscura/ObscuraService.java b/services/core/java/com/android/server/obscura/ObscuraService.java
index d31a504e93d9..fc9f49b9b4ac 100644
--- a/services/core/java/com/android/server/obscura/ObscuraService.java
+++ b/services/core/java/com/android/server/obscura/ObscuraService.java
@@ -265,5 +265,9 @@ public class ObscuraService extends IObscuraManager.Stub implements IObscuraServ
         return SettingsSpoofController.getSpoofedValue(settingName);
     }
 
+    @Override
+    public void launchHiddenApp(String packageName) {
+        mAppControlController.launchHiddenApp(packageName);
+    }
 
 }
